crates/turbopack-ecmascript/src/analyzer/mod.rs

```diff
@@ -147,60 +180,131 @@ pub struct ModuleValue {
+    /// might contain placeholders.
+    Operation,
+    /// These values are replaced during linking.
+    Placeholder,
+}
+
 /// TODO: Use `Arc`
+/// There are 4 kinds of values: Leafs, Nested, Operations, and Placeholders
+/// (see [JsValueMetaKind] for details). Values are processed in two phases:
+/// - Analyse phase: We convert AST into [JsValue]s. We don't have contextual
+///   information so we need to insert placeholders to represent that.
+/// - Link phase: We try to reduce a value to a constant value.
+/// The link phase as 5 substeps that are executed on each node in the graph
+/// depth-first. When a value was modified, we need to visit the new children
+/// again.
+/// - Replace variables with their values. This replaces [JsValue::Variable]. No
+///   variables should be remaining after that.
+/// - Replace placeholders with contextual information. This usually replaces
+///   [JsValue::FreeVar] and [JsValue::Module]. Some [JsValue::Call] on well
+///   known functions might also be replaced. No free vars or modules should be
+///   remaining after that.
+/// - Replace operations on well-known objects and functions. THis handles
+///   [JsValue::Call] and [JsValue::Member] on well-known objects and functions.
+/// - Replace all built-in functions with their values when they are
+///   compile-time constant.
+/// - For optimization any nested operations are replaced with
+///   [JsValue::Unknown]. So only one layer of operation remains.
+/// Any remaining operation or placeholder can be treated as unknown.
 #[derive(Debug, Clone, Hash, PartialEq, Eq)]
 pub enum JsValue {
-    /// Denotes a single string literal, which does not have any unknown value.
-    ///
-    /// TODO: Use a type without span
+    // LEAF VALUES
+    // ----------------------------
+    /// A constant primitive value.
     Constant(ConstantValue),
+    /// An constant URL object.
+    Url(Url),
+    /// Some kind of well known object
+    /// (must not be an array, otherwise Array.concat need to be changed)
+    WellKnownObject(WellKnownObjectKind),
+    /// Some kind of well known function
+    WellKnownFunction(WellKnownFunctionKind),
+    /// Not analyzable value. Might contain the original value for additional
+    /// info. Has a reason string for explanation.
+    Unknown(Option<Arc<JsValue>>, &'static str),
 
+    // NESTED VALUES
+    // ----------------------------
+    /// An arrow of nested values
     Array(usize, Vec<JsValue>),
-
+    /// An object of nested values
     Object(usize, Vec<ObjectPart>),
-
-    Url(Url),
-
+    /// A list of alternative values
     Alternatives(usize, Vec<JsValue>),
-
-    // TODO no predefined kinds, only JsWord
-    FreeVar(FreeVarKind),
-
-    Variable(Id),
-
+    /// A function reference. The return value might contain [JsValue::Argument]
+    /// placeholders that need to be replaced when calling this function.
+    /// `(total_node_count, func_ident, return_value)`
+    Function(usize, u32, Box<JsValue>),
```

Comment:
Using struct-like variants is:
1. more robust; and
2. more readable; but
3. more verbose.

More verbose is an acceptable tradeoff.

You might be able to use https://github.com/rust-lang/rust-analyzer/pull/13303

```suggestion
    Function {
        total_node_count: usize,
        func_ident: u32,
        return_value: Box<JsValue>
    },
```

Reply:
follow-up PR